network/datadog_checks/network/network.py

```diff
@@ -298,7 +303,7 @@ def _check_linux(self, instance):
                     # bug that print `tcp` even if it's `udp`
                     # The `-H` flag isn't available on old versions of `ss`.
                     cmd = "ss --numeric --tcp --all --ipv{} | cut -d ' ' -f 1 | sort | uniq -c".format(ip_version)
-                    output, _, _ = get_subprocess_output(["sh", "-c", cmd], self.log)
+                    output, _, _ = get_subprocess_output(["sh", "-c", cmd], self.log, env={"PROC_ROOT": proc_location})
```

Comment:
is the env var correctly passed to `cmd`, even though it's executed by `sh -c`?

Reply:
also, do we need to pass `proc_location` or `net_proc_base_location`? (there's a subtle difference that I can't remember)

FWIW, `_is_collect_cx_state_runnable` is called with `net_proc_base_location`

Reply:
> is the env var correctly passed to `cmd`, even though it's executed by `sh -c` ?

**Yes**. I don’t think `sh -c` is supposed to touch the environment variables at all.
At least, it doesn’t in my test:

```shell
vagrant@agent-dev-lenaic:/tmp$ env -i FOO=bar sh -c 'env'
FOO=bar
PWD=/tmp
```

Things would have been different with `sh -l` (`--login`) or `sh -i` (`--interactive`) since a login (or interactive) shell would source shell specific scripts (`profile`, `bashrc`, `zshrc`, etc.):

```shell
$ env -i FOO=bar sh -c -l 'env'
PATH=/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
FOO=bar
LANG=C.UTF-8
GOPATH=/go
PWD=/tmp
LC_ALL=C.UTF-8
```

Reply:
> also, do we need to pass `proc_location` or `net_proc_base_location` ? (there's a subtle difference that I can't remember)

Good catch!

```python
    @staticmethod
    def _get_net_proc_base_location(proc_location):
        if Platform.is_containerized() and proc_location != "/proc":
            net_proc_base_location = "%s/1" % proc_location
        else:
            net_proc_base_location = proc_location
        return net_proc_base_location
```

So, here is the explanation of the trick:
Inside a container, the `procfs` of the host will be bind-mounted in, let’s say, `/host/proc`.
Network tools (including `ss`) are looking for network data inside `<procfs>/net` (where `<procfs>` is the value of the `PROC_ROOT` environment variable or `/proc` if it is not set).
But, `<procfs>/net` is a symlink to `self/net` so that the network tools report data for the network namespace they are running in.
In the agent case, the containerised agent is interested in the network data of the host and not of itself.
That’s why we want to use `/host/proc/1/net` (network data of the init process on the host) rather than `/host/proc/self/net` (network data of the agent network namespace).

So, it’s definitively `net_proc_base_location` that we need.

Démo:

On a host:
```
vagrant@agent-dev-lenaic:/tmp$ ss -tan
State     Recv-Q    Send-Q       Local Address:Port        Peer Address:Port
LISTEN    0         128              127.0.0.1:44097            0.0.0.0:*
LISTEN    0         64                 0.0.0.0:46637            0.0.0.0:*
LISTEN    0         128                0.0.0.0:111              0.0.0.0:*
LISTEN    0         128                0.0.0.0:46003            0.0.0.0:*
LISTEN    0         128          127.0.0.53%lo:53               0.0.0.0:*
LISTEN    0         128                0.0.0.0:22               0.0.0.0:*
ESTAB     0         0                10.0.2.15:22              10.0.2.2:50949
LISTEN    0         128                   [::]:111                 [::]:*
LISTEN    0         128                   [::]:35089               [::]:*
LISTEN    0         128                   [::]:22                  [::]:*
LISTEN    0         64                    [::]:37147               [::]:*
```

Inside a container with `proc_location`:
```
vagrant@agent-dev-lenaic:/tmp$ docker run -ti --rm -v /proc:/host/proc:ro debian env PROC_ROOT=/host/proc ss -tan
State    Recv-Q    Send-Q       Local Address:Port         Peer Address:Port
```

Inside a container with `net_proc_base_location`:
```
vagrant@agent-dev-lenaic:/tmp$ docker run -ti --rm -v /proc:/host/proc:ro debian env PROC_ROOT=/host/proc/1 ss -tan
State     Recv-Q    Send-Q       Local Address:Port        Peer Address:Port
LISTEN    0         0                127.0.0.1:44097            0.0.0.0:*
LISTEN    0         0                  0.0.0.0:46637            0.0.0.0:*
LISTEN    0         0                  0.0.0.0:111              0.0.0.0:*
LISTEN    0         0                  0.0.0.0:46003            0.0.0.0:*
LISTEN    0         0               127.0.0.53:53               0.0.0.0:*
LISTEN    0         0                  0.0.0.0:22               0.0.0.0:*
ESTAB     0         0                10.0.2.15:22              10.0.2.2:50949
LISTEN    0         0                        *:111                    *:*
LISTEN    0         0                        *:35089                  *:*
LISTEN    0         0                        *:22                     *:*
LISTEN    0         0                        *:37147                  *:*
```

Reply:
Thanks for the very detailed responses! 😄